extensions-core/hdfs-storage/src/main/java/org/apache/druid/inputsource/hdfs/HdfsInputSource.java

```diff
@@ -0,0 +1,238 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.inputsource.hdfs;
+
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.druid.data.input.AbstractInputSource;
+import org.apache.druid.data.input.InputFormat;
+import org.apache.druid.data.input.InputRowSchema;
+import org.apache.druid.data.input.InputSourceReader;
+import org.apache.druid.data.input.InputSplit;
+import org.apache.druid.data.input.SplitHintSpec;
+import org.apache.druid.data.input.impl.InputEntityIteratingReader;
+import org.apache.druid.data.input.impl.SplittableInputSource;
+import org.apache.druid.java.util.common.IAE;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class HdfsInputSource extends AbstractInputSource implements SplittableInputSource<Path>
+{
+  private static final String PROP_PATHS = "paths";
+
+  private final List<String> inputPaths;
+  private final Configuration configuration;
+
+  @JsonCreator
+  public HdfsInputSource(
+      @JsonProperty(PROP_PATHS) Object inputPaths,
+      @JacksonInject Configuration configuration
+  )
+  {
+    // Note: Currently commas or globs in InputPaths are not supported
```

Comment:
Is this a TODO thing? Then would you please an issue for it?

Reply:
I think it would be best to do this the same way that HdfsFirehoseFactory does it. It uses TextInputFormat, but that's just to keep the code short. It could have used FileInputFormat instead (in fact, the method it calls is from FileInputFormat and is not overridden by TextInputFormat). One way to do this is:

```java
    return (new FileInputFormat<Object, Object>()
    {
      @Override
      public RecordReader<Object, Object> createRecordReader(
          org.apache.hadoop.mapreduce.InputSplit inputSplit,
          TaskAttemptContext taskAttemptContext
      )
      {
        throw new UnsupportedOperationException();
      }

      @Override
      protected boolean isSplitable(JobContext context, Path filename)
      {
        return false;
      }
    }).getSplits(job)
      .stream()
      .map(split -> ((FileSplit) split).getPath())
      .collect(Collectors.toSet());
```

Actually, this'd be somewhat better, since returning `false` from `isSplitable` prevents generating a ton of extra paths. (They're being deduped by the Set Collector, so they don't affect correctness, but they do affect memory use.)

Whatever you do, please share the code between HdfsFirehoseFactory and HdfsInputSource, unless there is a good reason not to. Maybe do a static method in HdfsInputSource and have HdfsFirehoseFactory call that. We'll want to remove the FirehoseFactories in the future, and that makes the future deletion easier.

Reply:
The javadoc for `SplittableInputSource` state that `createSplits()` should do lazy evaluations. When I looked at the implementation of `FileInputFormat.getSplits()`, my understanding is that it eagerly gets all of the values. Lazily evaluating the glob patterns looks a bit tricky, so for the first implementation of `HdfsInputSource`, I was thinking of just handling the non-glob case.

Reply:
That's true, but for these reasons, I think it makes sense to do it eagerly for now, with the Hadoop FS code approach:

1. It will improve compatibility with the `index_hadoop` task. It would be best if people can easily migrate from Hadoop.
2. `createSplits()` is specced to encourage laziness for memory usage reasons. But, the `index_hadoop` task does it eagerly, so this won't be an issue for anyone migrating from Hadoop.